### PR TITLE
Use any reward model for online methods

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

This PR allows any reward model to be used with Online DPO, i.e. it removes the requirement to have the same chat template and tokenizer.

The user must now provide `reward_processing_class`.
```python
trainer = OnlineDPOTrainer(
    model=model,
    reward_model=reward_model,
    args=training_args,
    train_dataset=train_dataset,
    processing_class=tokenizer,
    reward_processing_class=reward_tokenizer,  # <-
)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/online_dpo_trainer.py [hunk 1]
```diff
@@ -428,18 +430,23 @@ def training_step(
             ref_logprobs = torch.take_along_dim(ref_all_logprobs, completion_ids.unsqueeze(-1), dim=2).squeeze(-1)
             del ref_output, ref_logits, ref_all_logprobs  # free memory
 
-        # Get the reward from the reward model or judge:
-        if self.judge is not None:
-            completions = self.processing_class.batch_decode(
-                prompt_completion_ids[:, context_length:], skip_special_tokens=True
-            )
-            completions = [completion.strip() for completion in completions]  # remove the leading space
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:436]: I think we don't need strip

**Comment by qgallouedec** [comment]
Results for a gemma reward model

```
accelerate launch examples/scripts/dpo_online.py \
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
    --reward_model_path Ray2333/GRM-Gemma-2B-rewardmodel-ft \
    --dataset_name trl-lib/ultrafeedback-prompt \
    --learning_rate 5.0e-7 \
    --logging_steps 10 \
    --output_dir Qwen2-0.5B-OnlineDPO-GRM-Gemma \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 2 \
    --warmup_ratio 0.1 \
    --missing_eos_penalty 1.0 \
    --push_to_hub
```

https://wandb.ai/huggingface/huggingface/runs/520cnnjl

For ref, with Pair RM judge instead:

```
accelerate launch examples/scripts/dpo_online.py \
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
    --judge pair_rm \
    --dataset_name trl-lib/ultrafeedback-prompt \
    --learning_rate 5.0e-7 \
    --logging_steps 10 \
    --output_dir Qwen2-0.5B-OnlineDPO-PairRM \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 2 \
    --warmup_ratio 0.1 \
    --push_to_hub
```


https://wandb.ai/huggingface/huggingface/runs/ffd4u5wa

File: trl/trainer/online_dpo_trainer.py [hunk 2]
```diff
@@ -428,18 +430,23 @@ def training_step(
...
+        # Decode the completions, and format them if the input is conversational
+        device = prompt_completion_ids.device
+        completions_ids = prompt_completion_ids[:, context_length:]
+        completions = self.processing_class.batch_decode(completions_ids, skip_special_tokens=True)
+        if is_conversational({"prompt": prompts[0]}):
+            completions = [[{"role": "assistant", "content": completion}] for completion in completions]
 
+        # Get the reward from the reward model or judge
+        if self.judge is not None:
+            # Once formatted, conversational data may contain special tokens (such as <|im_start|>) that are not
```
Comment (lewtun) [inline trl/trainer/online_dpo_trainer.py:442]: I think one edge case here is when the prompt consists of multi-turn messages, but I think we can ignore this for now.

**Review by lewtun — APPROVED** [review]
Awesome PR @qgallouedec - this will unlock so many combinations of policy with models on RewardBench 🔥 !

Have you done a test run of e.g. trying to optimise Qwen2.5-0.5B-Instruct with the 7B ArmoRM model?

File: trl/trainer/online_dpo_trainer.py [hunk 3]
```diff
@@ -449,24 +456,47 @@ def training_step(
...
             # when rank == 1, it means the second completion is the best
-            mask = torch.tensor([rank == 0 for rank in ranks_of_first_completion], device=prompt_completion_ids.device)
+            mask = torch.tensor([rank == 0 for rank in ranks_of_first_completion], device=device)
         else:
-            _, scores, _ = get_reward(
-                self.reward_model, prompt_completion_ids, self.processing_class.pad_token_id, context_length
-            )
+            # The reward model may not have the same chat template or tokenizer as the model, so we need to use the
+            # raw data (string), apply the chat template (if needed), and tokenize it with the reward processing class.
+            prompts = 2 * prompts  # repeat the prompt: [prompt0, prompt1] -> [prompt0, prompt1, prompt0, prompt1]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/online_dpo_trainer.py:463]: At this point, we have:

```python
prompts = ["What color is the sky?", "What's the capital of France?"]
completions = ["Blue", "Lyon", "Green", "Paris"]
```

and later, we need to concat the prompts and the completions to compute the reward

```python
prompt_completion_ids = torch.cat((prompts_ids, completions_ids), dim=1)
_, scores, _ = get_reward(
    self.reward_model, prompt_completion_ids, self.reward_processing_class.pad_token_id, context_length
)
```

so we need to repeat the prompt.

File: docs/source/online_dpo_trainer.md [hunk 4]
```diff
@@ -79,20 +79,17 @@ Instead of a judge, you can chose to use a reward model -- see [Reward Bench](ht
 
 - judge = PairRMJudge()
 + reward_model = AutoModelForSequenceClassification.from_pretrained("trl-lib/Qwen2-0.5B-Reward", num_labels=1)
++ reward_tokenizer = AutoTokenizer.from_pretrained("trl-lib/Qwen2-0.5B-Reward")
 
   trainer = OnlineDPOTrainer(
       ...
 -     judge=judge,
 +     reward_model=reward_model,
++     reward_processing_class=reward_tokenizer,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply docs/source/online_dpo_trainer.md:88]: Possibly. And since the `tokenizer` is now called `processing_class` within trainers, I'd recommend always aligning with it (even if only the textual modality is supported). Unless you have a good reason not to.

File: examples/scripts/dpo_online.py [hunk 5]
```diff
@@ -93,8 +93,13 @@
             trust_remote_code=model_config.trust_remote_code,
             **model_kwargs,
         )
+        reward_tokenizer = AutoTokenizer.from_pretrained(
+            training_args.reward_model_path,
+            trust_remote_code=model_config.trust_remote_code,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/scripts/dpo_online.py:98]: Ok for truncation, and truncation side. Not sure what's the best way to let the user set the max_length. Ok for doing this in a follow-up PR?

File: trl/trainer/online_dpo_trainer.py (hunk 2 above)
Comment (qgallouedec) [reply trl/trainer/online_dpo_trainer.py:442]: Why would it be an edge-case?

File: examples/scripts/dpo_online.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/scripts/dpo_online.py:98]: 1834770

**Comment by qgallouedec** [comment]
> Have you done a test run of e.g. trying to optimise Qwen2.5-0.5B-Instruct with the 7B ArmoRM model?

ArmoRM is a custom classifier (its code for using it is not standard). So our `get_reward` function probably won't work for it. However, by modifying the code a little, I still manage to use it, and this is what I get:

https://wandb.ai/huggingface/huggingface/runs/merlfqgx (screenshot to come)

File: trl/trainer/online_dpo_trainer.py (hunk 2 above)
Comment (lewtun) [reply trl/trainer/online_dpo_trainer.py:442]: Ah I see: the simple chat template assigns alternating `User:` and `Assistant:` roles - ignore my comment!
